File: reference/glref/dag.py
"""Reference implementation of GLDAGPartition [Alg 5] of arXiv 2608.30945
(docs/paper_notes.md §9): the near-linear-time algorithm for ``k``-``T``-connected
DAGs, weighted or unweighted.

The DAG solver needs no FEAC machinery. Its precondition is [Lem 9.1
k-conn-dag]: a DAG is ``k``-``T``-connected iff every non-terminal has out-degree
``≥ k``. Each step contracts, into an active terminal ``t_i``, the
``≺``-earliest unused pre-terminal of the current part ``V_i`` (canonical
topological order [Def 9.2]); by [Lem 9.4 contract-exists-dag] it has no common
predecessor with ``t_i``, so the contraction keeps the contracted DAG
``k``-``T``-connected [Lem 9.3 dag-contract-theorem] and the parts satisfy
``Σ_{v∈V_i\\T} w_v ≤ c_i + w_max − 1`` [Lem 9.5 dag-algo-correctness].
"""
from __future__ import annotations

import heapq

from glsolver.instance import Instance

from .trace import Tracer
from .unweighted import InvariantError, RefResult, RefStats

POLICIES = ("max_residual", "round_robin", "first")


# ---------------------------------------------------------------------------
# structure tests
# ---------------------------------------------------------------------------


def _distinct_out_adjacency(inst: Instance) -> list[list[int]]:
    """Out-neighbour lists with duplicate arcs merged (first occurrence kept).

    The paper's digraphs are simple (paper_notes §2), so ``d^+(v) = |N^+(v)|``
    counts *distinct* out-neighbours. A normalized :class:`Instance` has no
    duplicate arcs; the structure tests below must nevertheless not mistake
    the arc multiplicity of a hand-built instance for its out-degree [Lem 9.1].
    """
    return [list(dict.fromkeys(nbrs)) for nbrs in inst.out_adjacency()]


def _kahn_nonterminals(inst: Instance) -> tuple[list[int], list[int]]:
    """Kahn's algorithm restricted to the non-terminals (arcs into terminals
    are irrelevant for acyclicity since terminals have no out-arcs).

    Returns ``(order, leftover)``: ``order`` is a topological order of the
    non-terminals that were processed (smallest available id first, for
    determinism) and ``leftover`` lists the non-terminals on or behind a cycle
    (empty iff acyclic). Duplicate arcs are merged first, so in-degrees and
    out-lists agree on distinct neighbours.
    """
    tset = set(inst.terminals)
    indeg = [0] * inst.n
    out = _distinct_out_adjacency(inst)
    for u in range(inst.n):
        for v in out[u]:
            if v not in tset:
                indeg[v] += 1
    ready = [v for v in range(inst.n) if v not in tset and indeg[v] == 0]
    heapq.heapify(ready)
    order: list[int] = []
    while ready:
        v = heapq.heappop(ready)
        order.append(v)
        for x in out[v]:
            if x in tset:
                continue
            indeg[x] -= 1
            if indeg[x] == 0:
                heapq.heappush(ready, x)
    leftover = [v for v in range(inst.n) if v not in tset and indeg[v] > 0]
    return order, leftover


def is_dag(inst: Instance) -> bool:
    """``True`` iff the arc set of ``inst`` is acyclic."""
    return not _kahn_nonterminals(inst)[1]


def canonical_topological_order(inst: Instance) -> list[int]:
    """[Def 9.2] canonical topological order: the non-terminals in a topological
    order (Kahn's algorithm, smallest available id first) followed by the
    terminals in ``inst.terminals`` order. Raises ``ValueError`` on a cycle."""
    order, leftover = _kahn_nonterminals(inst)
    if leftover:
        raise ValueError(f"graph has a directed cycle through {leftover[:5]}; no topological order")
    return order + list(inst.terminals)


def _precondition_detail(inst: Instance) -> tuple[bool, int | None, str]:
    """[Lem 9.1] on the *distinct* out-neighbour sets, then acyclicity.

    Well-formedness (no duplicate arcs, no self-loops, ...) is not checked
    here; :func:`gl_dag_partition` calls ``inst.validate()`` for that.
    """
    k = inst.k
    tset = set(inst.terminals)
    out = _distinct_out_adjacency(inst)  # d^+(v) = |N^+(v)|, not the arc multiplicity
    for v in range(inst.n):
        if v in tset:
            continue
        if len(out[v]) < k:
            return False, v, (
                f"non-terminal {v} has out-degree {len(out[v])} < k = {k}; "
                f"the DAG is not k-T-connected [Lem 9.1]"
            )
    _order, leftover = _kahn_nonterminals(inst)
    if leftover:
        v = leftover[0]
        return False, v, f"graph is not acyclic: vertex {v} lies on or behind a directed cycle"
    return True, None, "ok"


def dag_precondition(inst: Instance) -> tuple[bool, int | None]:
    """[Lem 9.1] precondition of [Alg 5]: ``inst`` is a DAG **and** every
    non-terminal has out-degree ``≥ k``. Returns ``(ok, offending_vertex)``
    where the vertex is a non-terminal of out-degree ``< k`` or a vertex on a
    cycle (``None`` when ``ok``).

    Out-degree means the number of *distinct* out-neighbours (paper_notes §2),
    so a hand-built instance with duplicate arcs is judged like its normalized
    form. This is a pure query and never validates the instance; malformed
    input is rejected by :func:`gl_dag_partition` (``inst.validate()``)."""
    ok, v, _why = _precondition_detail(inst)
    return ok, v


def is_k_t_connected_dag(inst: Instance) -> bool:
    """[Lem 9.1 k-conn-dag] (name used in the paper_notes §12 lemma map)."""
    return dag_precondition(inst)[0]


# ---------------------------------------------------------------------------
# [Alg 5] GLDAGPartition
# ---------------------------------------------------------------------------


def gl_dag_partition(
    inst: Instance,
    *,
    policy: str = "max_residual",
    tracer: Tracer | None = None,
    debug: bool = False,
) -> RefResult:
    """[Alg 5] GLDAGPartition (paper_notes §9).

    Per terminal ``t_i`` a min-heap ``H_i`` keyed by the position in the
    canonical topological order ``≺`` holds entries ``(pos[p], p, x)`` where
    ``x ∈ V_i`` is the head of the arc ``(p, x)`` that put ``p`` into the heap
    (so ``parents[p] = x`` is an original arc into the part, §9). Stale entries
    (``used[p]``) are skipped on extraction. ``policy`` chooses the active
    terminal (§13.7): ``"max_residual"`` (largest residual capacity, ties by
    terminal index), ``"round_robin"`` or ``"first"`` (smallest index); every
    policy is valid. Weights are supported: the residual capacity drops by
    ``w_p`` and the terminal becomes inactive once it is ``≤ 0``.

    Raises ``ValueError`` for an unknown ``policy`` or a structurally
    malformed instance (``inst.validate()``: duplicate arcs, self-loops, arcs
    leaving a terminal, wrong capacity/weight sums), so such input never
    reaches the heap loop. Returns ``status == "precondition_failed"`` when
    [Lem 9.1] fails (some out-degree ``< k`` or a cycle) with the offending
    vertex in ``message``.
    Debug mode checks, before each contraction, the hypothesis of [Lem 9.3]:
    ``p`` and ``t_i`` have no common predecessor among the unused vertices
    ([Lem 9.4]; ``O(in-degree × |out|)`` per step) and, after it, that no
    unused non-terminal lost out-degree in the contracted graph (simulated).
    """
    if policy not in POLICIES:
        raise ValueError(f"unknown policy {policy!r}; expected one of {POLICIES}")
    inst.validate()  # well-formedness (duplicate arcs, self-loops, sums, ...): ValueError
    tracer = tracer or Tracer(enabled=False)
    stats = RefStats()
    ok, _bad, why = _precondition_detail(inst)
    if not ok:
        return RefResult("precondition_failed", [], {}, {}, stats, why, tracer.events if tracer.enabled else None)

    n, k = inst.n, inst.k
    terminals = list(inst.terminals)
    tset = set(terminals)
    order = canonical_topological_order(inst)
    pos = {v: i for i, v in enumerate(order)}
    in_adj = inst.in_adjacency()
    out_adj = inst.out_adjacency()
    if inst.weights is None:
        w = [1] * n
    else:
        w = [int(x) for x in inst.weights]
    residual = list(inst.capacities)
    parts: list[list[int]] = [[t] for t in terminals]
    part_of: list[int] = [-1] * n  # index of the part a used vertex (or terminal) belongs to
    for i, t in enumerate(terminals):
        part_of[t] = i
    used = [False] * n
    heaps: list[list[tuple[int, int, int]]] = []
    for t in terminals:
        h = [(pos[u], u, t) for u in in_adj[t]]
        heapq.heapify(h)
        stats.heap_pushes += len(h)
        heaps.append(h)
    active = [i for i in range(k) if residual[i] > 0]
    rr = 0
    r = n - k
    parents: dict[int, int] = {}

    tracer.record(
        "init", n=n, k=k, terminals=terminals, capacities=list(inst.capacities),
        arcs=[list(a) for a in inst.arcs], weights=list(w), order=order, policy=policy,
    )

    def _check_out_degrees() -> None:
        """Debug: every unused non-terminal keeps out-degree >= k in the contracted graph."""
        for u in range(n):
            if u in tset or used[u]:
                continue
            heads: set[int] = set()
            for y in out_adj[u]:
                if used[y] or y in tset:
                    heads.add(terminals[part_of[y]])
                else:
                    heads.add(y)
            if len(heads) < k:
                raise InvariantError(
                    f"[Lem 9.3] vertex {u} has contracted out-degree {len(heads)} < k = {k}"
                )

    while r > 0:
        if not active:
            raise InvariantError("no active terminal although non-terminals remain (Σw ≤ Σc violated)")
        if policy == "first":
            i = active[0]
        elif policy == "round_robin":
            i = active[rr % len(active)]
            rr += 1
        else:  # max_residual: largest residual capacity, ties by terminal index
            i = min(active, key=lambda j: (-residual[j], j))
        H = heaps[i]
        p = -1
        x = -1
        while H:
            _, cand, head = heapq.heappop(H)
            stats.heap_pops += 1
            if used[cand]:
                stats.stale_pops += 1
                continue
            p, x = cand, head
            break
        if p < 0:
            raise InvariantError(f"[Lem 9.4] heap of active terminal {terminals[i]} ran dry")
        if debug:  # [Lem 9.4]: no unused common predecessor of p and the part V_i
            for u in in_adj[p]:
                if used[u]:
                    continue
                if any((used[y] or y in tset) and part_of[y] == i for y in out_adj[u]):
                    raise InvariantError(
                        f"[Lem 9.4] {u} is a common predecessor of {p} and the part of {terminals[i]}"
                    )
        used[p] = True
        part_of[p] = i
        parts[i].append(p)
        parents[p] = x
        residual[i] -= w[p]
        r -= 1
        stats.contractions += 1
        for u in in_adj[p]:  # new pre-terminals of the enlarged part (stale if already used)
            heapq.heappush(H, (pos[u], u, p))
            stats.heap_pushes += 1
        tracer.record("dag_contract", p=p, t=terminals[i], parent=x, residual=residual[i])
        if residual[i] <= 0:
            active.remove(i)
        if debug:
            _check_out_degrees()

    stats.steps = stats.contractions
    tracer.record("done", parts={terminals[i]: sorted(parts[i]) for i in range(k)}, parents=dict(parents))
    return RefResult("ok", parts, parents, {}, stats, "ok", tracer.events if tracer.enabled else None)

File: src/glcore/dag.cpp
// GLDAGPartition [Alg 5] (docs/paper_notes.md §9) in two exactly equivalent implementations:
//   variant 0: one binary min-heap per terminal keyed by the canonical position [Def 9.2], O(m log n) [Lem 9.6];
//   variant 1: the stack-of-sorted-in-lists variant of RESEARCH_NOTES.md P1, O(n + m) (+ O(n log k) for the
//              max-residual policy).
// Both variants perform the same sequence of contractions and record the same parents (proof in P1), which
// tests/test_core_dag.py asserts against each other and against the reference glref.dag.gl_dag_partition.
#include "dag.hpp"

#include <algorithm>
#include <functional>
#include <queue>
#include <stdexcept>
#include <string>
#include <tuple>
#include <utility>
#include <vector>

namespace glcore {
namespace {

template <class T>
std::string vec_json(const std::vector<T>& v) {
    std::string s = "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i) s += ',';
        s += std::to_string(v[i]);
    }
    s += ']';
    return s;
}

// Heap entry of [Alg 5]: (pos[p], p, x) with x ∈ V_i the head of the arc (p, x) that created the entry. The
// lexicographic order is exactly the reference's heapq tuple order, so the first non-stale pop of p carries
// the smallest such x (paper_notes §9: "push pairs (pos[p], p, x); the first non-stale pop wins").
struct HeapEntry {
    int pos, p, x;
    bool operator>(const HeapEntry& o) const { return std::tie(pos, p, x) > std::tie(o.pos, o.p, o.x); }
};
using TermHeap = std::priority_queue<HeapEntry, std::vector<HeapEntry>, std::greater<HeapEntry>>;

// Stack level of RESEARCH_NOTES.md P1: a vertex x ∈ V_i and a cursor into its pos-sorted in-list.
struct Level {
    int x, ptr;
};

// Key of the max-residual policy queue: largest residual first, ties by smallest terminal index (§13.7).
struct ResidualKey {
    int64_t residual;
    int idx;
    bool operator<(const ResidualKey& o) const {  // "less" = lower priority
        if (residual != o.residual) return residual < o.residual;
        return idx > o.idx;
    }
};

DagResult fail(const std::string& msg) { return DagResult{"precondition_failed", msg, {}, {}}; }

}  // namespace

// [Alg 5] GLDAGPartition. See dag.hpp for the parameter contract.
DagResult dag_partition(int n, const std::vector<std::pair<int, int>>& arcs, const std::vector<int>& terminals,
                        const std::vector<int64_t>& capacities, const std::vector<int64_t>& weights,
                        int policy, int variant, Stats& stats, Trace& trace, bool check_precondition) {
    Stats::Timer timer(stats, "dag");
    // ------------------------------------------------------------------ input validation (malformed => throw)
    if (n <= 0) throw std::invalid_argument("dag_partition: n must be positive");
    const int k = (int)terminals.size();
    if (k <= 0) throw std::invalid_argument("dag_partition: at least one terminal is required");
    if ((int)capacities.size() != k) throw std::invalid_argument("dag_partition: capacities must have one entry per terminal");
    if (!weights.empty() && (int)weights.size() != n) throw std::invalid_argument("dag_partition: weights must be empty or have length n");
    if (policy < 0 || policy > 2) throw std::invalid_argument("dag_partition: policy must be 0 (max residual), 1 (round robin) or 2 (first)");
    if (variant < 0 || variant > 1) throw std::invalid_argument("dag_partition: variant must be 0 (heap) or 1 (linear stack)");
    std::vector<int> term_index(n, -1);
    for (int i = 0; i < k; ++i) {
        const int t = terminals[i];
        if (t < 0 || t >= n) throw std::invalid_argument("dag_partition: terminal " + std::to_string(t) + " out of range");
        if (term_index[t] >= 0) throw std::invalid_argument("dag_partition: terminal " + std::to_string(t) + " listed twice");
        term_index[t] = i;
        if (capacities[i] < 0) throw std::invalid_argument("dag_partition: capacity of terminal " + std::to_string(t) + " is negative");
    }
    for (int v = 0; v < n; ++v)
        if (!weights.empty() && term_index[v] < 0 && weights[v] < 1)
            throw std::invalid_argument("dag_partition: non-terminal " + std::to_string(v) + " must have weight >= 1");

    // ------------------------------------------------------------------ out-CSR (paper §2 conventions: drop
    // arcs leaving a terminal, self-loops and duplicate arcs; the Python layer normalizes already).
    std::vector<int> out_start(n + 1, 0), out_deg(n, 0);
    for (const auto& a : arcs) {
        const int u = a.first, v = a.second;
        if (u < 0 || u >= n || v < 0 || v >= n)
            throw std::invalid_argument("dag_partition: arc (" + std::to_string(u) + "," + std::to_string(v) + ") out of range");
        if (u == v || term_index[u] >= 0) continue;
        ++out_start[u + 1];
    }
    for (int v = 0; v < n; ++v) out_start[v + 1] += out_start[v];
    std::vector<int> out_list(out_start[n]);
    {
        std::vector<int> fill(out_start.begin(), out_start.end() - 1);
        for (const auto& a : arcs) {
            const int u = a.first, v = a.second;
            if (u == v || term_index[u] >= 0) continue;
            out_list[fill[u]++] = v;
        }
        // duplicate merge with a stamp array: O(n + m), order of first occurrence kept.
        std::vector<int> stamp(n, -1);
        for (int u = 0; u < n; ++u) {
            int w = out_start[u];
            for (int p = out_start[u]; p < out_start[u + 1]; ++p) {
                const int v = out_list[p];
                if (stamp[v] == u) continue;
                stamp[v] = u;
                out_list[w++] = v;
            }
            out_deg[u] = w - out_start[u];
        }
    }
    auto out_begin = [&](int u) { return out_start[u]; };
    auto out_end = [&](int u) { return out_start[u] + out_deg[u]; };

    // ------------------------------------------------------------------ precondition [Lem 9.1]: out-degree >= k
    if (check_precondition) {
        for (int v = 0; v < n; ++v) {
            if (term_index[v] >= 0) continue;
            if (out_deg[v] < k)
                return fail("non-terminal " + std::to_string(v) + " has out-degree " + std::to_string(out_deg[v]) + " < k = " +
                            std::to_string(k) + "; the DAG is not k-T-connected [Lem 9.1]");
        }
    }

    // ------------------------------------------------------------------ canonical topological order [Def 9.2]:
    // Kahn's algorithm on the non-terminals with a min-heap on vertex id (deterministic), then the terminals.
    std::vector<int> indeg(n, 0);
    for (int u = 0; u < n; ++u)
        for (int p = out_begin(u); p < out_end(u); ++p)
            if (term_index[out_list[p]] < 0) ++indeg[out_list[p]];
    std::vector<int> order;
    order.reserve(n);
    {
        std::priority_queue<int, std::vector<int>, std::greater<int>> ready;
        for (int v = 0; v < n; ++v)
            if (term_index[v] < 0 && indeg[v] == 0) ready.push(v);
        while (!ready.empty()) {
            const int v = ready.top();
            ready.pop();
            order.push_back(v);
            for (int p = out_begin(v); p < out_end(v); ++p) {
                const int x = out_list[p];
                if (term_index[x] >= 0) continue;
                if (--indeg[x] == 0) ready.push(x);
            }
        }
    }
    if ((int)order.size() != n - k) {
        for (int v = 0; v < n; ++v)
            if (term_index[v] < 0 && indeg[v] > 0)
                return fail("graph is not acyclic: vertex " + std::to_string(v) + " lies on or behind a directed cycle");
        throw std::logic_error("dag_partition: Kahn's algorithm lost vertices without a cycle");
    }
    for (int i = 0; i < k; ++i) order.push_back(terminals[i]);
    std::vector<int> pos(n);
    for (int i = 0; i < n; ++i) pos[order[i]] = i;

    // ------------------------------------------------------------------ Σ w_v <= Σ c_t (paper_notes §9 precondition)
    int64_t total_w = 0, total_c = 0;
    for (int v = 0; v < n; ++v)
        if (term_index[v] < 0) total_w += weights.empty() ? 1 : weights[v];
    for (int i = 0; i < k; ++i) total_c += capacities[i];
    if (total_w > total_c)
        return fail("total non-terminal weight " + std::to_string(total_w) + " exceeds total capacity " + std::to_string(total_c) +
                    " (Σ w_v <= Σ c_t required, paper_notes §9)");

    // ------------------------------------------------------------------ in-lists sorted by canonical position:
    // a global counting sort (iterate tails in ≺-order), O(n + m). Variant 0 does not need the order but
    // shares the structure.
    std::vector<int> in_start(n + 1, 0);
    for (int u = 0; u < n; ++u)
        for (int p = out_begin(u); p < out_end(u); ++p) ++in_start[out_list[p] + 1];
    for (int v = 0; v < n; ++v) in_start[v + 1] += in_start[v];
    std::vector<int> in_list(in_start[n]);
    {
        std::vector<int> fill(in_start.begin(), in_start.end() - 1);
        for (int i = 0; i < n - k; ++i) {  // non-terminals in ≺-order (terminals have no out-arcs)
            const int u = order[i];
            for (int p = out_begin(u); p < out_end(u); ++p) in_list[fill[out_list[p]]++] = u;
        }
    }

    // ------------------------------------------------------------------ state of [Alg 5]
    std::vector<char> used(n, 0);
    std::vector<int> part_of(n, -1), parent(n, -1);
    std::vector<int64_t> residual(capacities);
    for (int i = 0; i < k; ++i) part_of[terminals[i]] = i;
    std::vector<int> active;  // ascending terminal indices (policies 1 and 2 index into it like the reference)
    std::vector<char> is_active(k, 0);
    std::priority_queue<ResidualKey> res_pq;  // policy 0, lazy deletion (stale = residual changed / inactive)
    for (int i = 0; i < k; ++i) {
        if (residual[i] > 0) {
            active.push_back(i);
            is_active[i] = 1;
            if (policy == 0) res_pq.push(ResidualKey{residual[i], i});
        }
    }
    int64_t rr = 0;  // round-robin counter (reference semantics: active[rr % len(active)], rr += 1)
    int r = n - k;
    int64_t done = 0;  // contractions performed by this call

    std::vector<TermHeap> heaps;
    std::vector<std::vector<Level>> stacks;
    if (variant == 0) {
        heaps.resize(k);
        for (int i = 0; i < k; ++i) {
            const int t = terminals[i];
            for (int p = in_start[t]; p < in_start[t + 1]; ++p) heaps[i].push(HeapEntry{pos[in_list[p]], in_list[p], t});
        }
    } else {
        stacks.resize(k);
        for (int i = 0; i < k; ++i) stacks[i].push_back(Level{terminals[i], in_start[terminals[i]]});
    }

    if (trace.enabled) {
        std::string arcs_json = "[";
        bool first = true;
        for (int u = 0; u < n; ++u)
            for (int p = out_begin(u); p < out_end(u); ++p) {
                if (!first) arcs_json += ',';
                first = false;
                arcs_json += "[" + std::to_string(u) + "," + std::to_string(out_list[p]) + "]";
            }
        arcs_json += ']';
        std::vector<int64_t> w_out(n, 0);
        for (int v = 0; v < n; ++v)
            if (term_index[v] < 0) w_out[v] = weights.empty() ? 1 : weights[v];
        trace.record("init", "{\"n\":" + std::to_string(n) + ",\"k\":" + std::to_string(k) + ",\"terminals\":" + vec_json(terminals) +
                                 ",\"capacities\":" + vec_json(capacities) + ",\"arcs\":" + arcs_json + ",\"weights\":" + vec_json(w_out) +
                                 ",\"order\":" + vec_json(order) + ",\"policy\":" + std::to_string(policy) + ",\"variant\":" +
                                 std::to_string(variant) + "}");
    }

    // choose the active terminal (§13.7); every policy is valid.
    auto select_terminal = [&]() -> int {
        if (active.empty()) throw std::runtime_error("dag_partition: no active terminal although non-terminals remain (Σ w <= Σ c violated)");
        if (policy == 2) return active[0];
        if (policy == 1) {
            const int i = active[(size_t)(rr % (int64_t)active.size())];
            ++rr;
            return i;
        }
        while (!res_pq.empty()) {
            const ResidualKey key = res_pq.top();
            if (is_active[key.idx] && key.residual == residual[key.idx]) return key.idx;
            res_pq.pop();  // stale
        }
        throw std::logic_error("dag_partition: residual queue empty while terminals are active");
    };
    auto deactivate = [&](int i) {
        is_active[i] = 0;
        active.erase(std::find(active.begin(), active.end(), i));
    };

    // ------------------------------------------------------------------ main loop of [Alg 5]
    while (r > 0) {
        const int i = select_terminal();
        int p = -1, x = -1;
        if (variant == 0) {
            // extract-min, skipping stale entries [Alg 5]; the entry's x is the parent (min over V_i by order).
            TermHeap& H = heaps[i];
            while (!H.empty()) {
                const HeapEntry e = H.top();
                H.pop();
                if (used[e.p]) continue;
                p = e.p;
                x = e.x;
                break;
            }
        } else {
            // P1: the minimum unused entry of H_i is the first unused element of the deepest non-exhausted level.
            std::vector<Level>& S = stacks[i];
            while (!S.empty()) {
                Level& L = S.back();
                const int end = in_start[L.x + 1];
                while (L.ptr < end && used[in_list[L.ptr]]) ++L.ptr;
                if (L.ptr == end) {
                    S.pop_back();
                    continue;
                }
                p = in_list[L.ptr++];
                break;
            }
            if (p >= 0) {
                // parent = smallest x ∈ V_i with (p, x) ∈ E: exactly the x of the minimum heap entry of p.
                for (int q = out_begin(p); q < out_end(p); ++q) {
                    const int y = out_list[q];
                    if (part_of[y] == i && (x < 0 || y < x)) x = y;
                }
                if (x < 0) throw std::logic_error("dag_partition: stack variant selected a vertex without an arc into the part");
            }
        }
        if (p < 0)
            throw std::runtime_error("[Lem 9.4] heap of active terminal " + std::to_string(terminals[i]) +
                                     " ran dry; the input is not k-T-connected");
        // contract p into t_i [Lem 9.3]: V_i += {p}, residual -= w_p, new pre-terminals = in-neighbours of p.
        used[p] = 1;
        part_of[p] = i;
        parent[p] = x;
        residual[i] -= weights.empty() ? 1 : weights[p];
        --r;
        ++done;
        if (variant == 0) {
            TermHeap& H = heaps[i];
            for (int q = in_start[p]; q < in_start[p + 1]; ++q) {
                const int u = in_list[q];
                if (!used[u]) H.push(HeapEntry{pos[u], u, p});  // used u would be a stale entry anyway
            }
        } else {
            stacks[i].push_back(Level{p, in_start[p]});
        }
        if (trace.enabled)
            trace.record("dag_contract", "{\"p\":" + std::to_string(p) + ",\"t\":" + std::to_string(terminals[i]) + ",\"parent\":" +
                                             std::to_string(x) + ",\"residual\":" + std::to_string(residual[i]) + "}");
        if (residual[i] <= 0) {
            deactivate(i);
        } else if (policy == 0) {
            res_pq.push(ResidualKey{residual[i], i});
        }
    }
    stats.contractions += done;
    stats.steps += done;

    if (trace.enabled) {
        std::vector<std::vector<int>> parts(k);
        for (int v = 0; v < n; ++v) parts[part_of[v]].push_back(v);
        std::string s = "{\"parts\":{";
        for (int i = 0; i < k; ++i) {
            if (i) s += ',';
            s += "\"" + std::to_string(terminals[i]) + "\":" + vec_json(parts[i]);
        }
        s += "}}";
        trace.record("done", s);
    }
    return DagResult{"ok", "ok", std::move(part_of), std::move(parent)};
}

}  // namespace glcore
